Pm interface refactor

jd-lara opened the pull request:
This PR updates the interface of PowerSimulations with PowerModels to address the requirements of some users to have access to the flow variables.  

In Summary. 

- Creates a PowerSimulations reduced version of `core_ref!` 
- Substitutes `build_generic_model` with a model instantiation 
- Changes the creation of the line flow variables in order to reference them in the PowerSimulations model and the PowerModels model. 
- implements #105 partially for some variables
- Updates the testing  

closes #148

ccoffrin commented:
I will have a chance to review this in the next day or so.

ccoffrin commented on src/devices/device_models/AC_branches.jl:
```diff
@@ -26,89 +26,252 @@ function flow_variables(ps_m::CanonicalModel,
+    #Get PowerModels dicts
+    pm_object = ps_m.pm_model
+    ref = pm_object.ref
+    var = pm_object.var
```
Instead of accessing the internal ref and var objects.  I recomend to use the ref and var functions (e.g. `var(pm_object, t, :p)`).  These allow you to ignore more advanced PowerModels features, like conductors.

jd-lara replied:
solved via b9d9919

ccoffrin commented on src/devices/device_models/AC_branches.jl:
```diff
@@ -26,89 +26,252 @@ function flow_variables(ps_m::CanonicalModel,
+                                                                            base_name="$(bus_fr),$(bus_to)_{$(d.name),$(t)}",
+                                                                            upper_bound = d.rate,
+                                                                            lower_bound = -d.rate,
+                                                                            start = 0.0)
```
Instead of creating these variables yourself, why not let PowerModels create them and extract them into your data structure?

ccoffrin commented on src/devices/device_models/common/pm_translator.jl:
```diff
@@ -143,13 +143,19 @@ function get_branches_to_pm(sys::PSY.System)
+            "bus_type" => 2
+        elseif bus.bustype == "SF"
+            "bus_type" => 3
+        end,
```
Missing type 4?

jd-lara replied:
We don't have islanded buses in PowerSystems.Bus currently.

ccoffrin commented on src/network/network_models/powermodels_interface.jl:
```diff
@@ -62,17 +144,14 @@ end
     for (n, network) in PM.nws(pm)
         @assert !PM.ismulticonductor(pm, nw=n)
         PM.variable_voltage(pm, nw=n)
-        PM.variable_branch_flow(pm, nw=n)#, bounded=false)
```
I would not recomend removing this.

ccoffrin commented on src/network/network_models/powermodels_interface.jl:
```diff
@@ -255,20 +334,26 @@ function powermodels_network!(ps_m::CanonicalModel,
-        pm_data["nw"]["$(t)"]["bus"]["$(bus.number)"]["qni"] = ps_m.expressions[:nodal_balance_reactive][bus.number,t]
+    for t in time_steps
+        var[:nw][t][:cnd][1][:p] = JuMP.Containers.DenseAxisArray(var[:nw][t][:cnd][1][:p],ref[:nw][t][:arcs])
+        var[:nw][t][:cnd][1][:q] = JuMP.Containers.DenseAxisArray(var[:nw][t][:cnd][1][:q],ref[:nw][t][:arcs])
```
It is not clear to me why this mapping is happening here.

jd-lara commented:
The objective of this PR is to create a way to make the branch flow constraints in PowerSimulations such that we can impose other constraints like flow schedules in interties or make branch models for controllable branches like tap changer transformers or angle dispatch with phase shifters. The process to integrate the constraints and variables from PowerModels into PowerSimulations follows this procedure in the current version of v0.2.0. 

1. We load the data from the System struct into a PowerModels dict, and we loop over the branches once here. 
2. We then call the `GenericPowerModel` with the formulation spec and returns an anonymous function we use in the `build_generic_model` function call. 
3. `build_generic_model` runs `core_ref!` and goes over the data again to makes the ref fields. After the `build_generic_model` call the `post_nip_expr method` that goes over the data again to make the variables. 
4. *If we reference the variables in PowerSimulations, we would go over the branches and time_steps once more and loop over them again impose further constraints if we need it.  

The idea is to take the previous process and split it in two to integrate it better as follows: 

1. Remove the anonymous function creation and instantiate the PowerModels object with the buses only in the `canonical_model` build,  add the place holder for the variables and constraints similar to what the `GenericPowerModels` function call does. Here part a part of `core_ref!` got added such that there is a place holder for the archs when they get added during the variable creation. Maybe this is not the best place to do it, but I need the keys in the `ref` dict before going into the variable creation. 
2. When we get to creating the device models, the code loops over the branches only once. In that loop the data is added to the `ref` dict directly, generate the arcs for the constraint creation, create the variables and reference them both in the PowerSimulations `canonical_model` and in the PowerModels var dict that was pre-generated in step 1. The idea is that by doing this, we can avoid step 4. 
3. Once this is done, call the remaining routines of `core_ref!` (a simplified version is what I added in `psi_ref!` that is needed for the integration and run the `post_nip_expr` but now the flow variables have been already created, and the arcs added accordingly. 

@ccoffrin is correct that this exposes a lot of the PowerModels complexity in the `add_flow_variable` calls and it is not super clean. The idea is to remove the repeated loops over the branches data and variables.

ccoffrin commented:
A few more questions based on your response.

Re flow schedules on interties:  Why not do the following, let power models build the flow variables, then you impose extra constraints on them?

Re tap changer transformers or angle dispatch with phase shifters: I would recommend we just add this type of component as a feature to PowerModels, which you can then use through the data model.  Based on the types of problems you have discussed in the past, I highly doubt PowerSimulations wants to get into the complexities of modeling tap changers; It gets particularly complex once you consider variants of complex power beyond AC Polar.  Phase shifters in the DC model are fairly easy to add some some models (e.g. DC), but following the previous devision of functionality, this would be best implemented in PowerModels.

jd-lara commented:
> Re flow schedules on interties: Why not do the following, let power models build the flow variables, then you impose extra constraints on them?

I wanted to avoid looping over the branches twice. This created some uncleanliness in the integration.  
 
> Re tap changer transformers or angle dispatch with phase shifters: I would recommend we just add this type of component as a feature to PowerModels, which you can then use through the data model. Based on the types of problems you have discussed in the past, I highly doubt PowerSimulations wants to get into the complexities of modeling tap changers; It gets particularly complex once you consider variants of complex power beyond AC Polar. Phase shifters in the DC model are fairly easy to add some some models (e.g. DC), but following the previous devision of functionality, this would be best implemented in PowerModels.

It's true it will be possible to contribute those models to PowerModels instead of making them ourselves. 

Overall I think @ccoffrin comments make sense, I can make a referencing using the accessor functions 
between the `canonical_model` and the `pm_model` to add more constraints to the flows when necessary. 

@claytonpbarrows if you are ok, I will close this PR and then try a new variant.

claytonpbarrows commented:
sounds good to me
